### PR TITLE
fix(skill-stocktake): inventory skills installed as symlinks (#2801)

### LINKED ISSUE #2801: skill-stocktake scan.sh misses symlinked skill directories
skills/skill-stocktake/scripts/scan.sh uses plain `find "$dir" -name "*.md" -type f`. Plain `find` doesn't descend into symlinked subdirectories, so any skill installed as a symlink (common: package-manager installs, app-bundled skills, cross-tool shared skills) is silently dropped from the inventory.

Repro: 4 skills under ~/.claude/skills, 3 of them symlinks. scan.sh reported only 1.

Fix: use `find -L "$dir" -name "*.md" -type f`.

### PR DESCRIPTION
**Pull request by ntdat812**
Fixes #2801.

## What Changed

`find "$dir" -name "*.md" -type f` → `find -L "$dir" -name "*.md" -type f`, in **both** `scan.sh` and `quick-diff.sh`.

## Why This Change

Plain `find` does not descend into a symlinked directory, and `-type f` is false for a symlinked file, so a skill installed as a link is invisible to the inventory. That covers the common cases: package-manager installs, app-bundled skills, and skills shared between tools. The reporter had 4 skills, 3 of them symlinks, and `scan.sh` reported 1.

The issue names only `scan.sh`. `quick-diff.sh:77` carries the identical expression, and it is the script that compares one inventory against another — fixing only `scan.sh` would have left them disagreeing, so every symlinked skill would show up as an addition or a removal on each run. Both are fixed here.

## Testing Done

- [x] Manual testing completed
- [ ] Automated tests pass locally (`node tests/run-all.js`) — see below
- [x] Edge cases considered and tested

**Measured against the real `scan.sh`**, with a fixture of 2 real and 2 symlinked skill directories, 4 `SKILL.md` on disk:

```
before:  scan.sh enumerated: 2 skill file(s)
after:   scan.sh enumerated: 4 skill file(s)
```

A note on how that was possible here: `ln -s` on this machine fails with `Operation not permitted` (no Developer Mode, not elevated), so the first attempt could not reproduce the bug at all — `ln -s` silently produced a directory **copy**, and both `find` forms then agreed. A Windows **directory junction** does not need elevation and Git Bash's `find` treats it as a symlink, which reproduces the defect exactly:

```
$ find .    -name "*.md" -type f     ->  ./real/a.md
$ find -L . -name "*.md" -type f     ->  ./linked/a.md  ./real/a.md
```

**New test: `tests/skills/skill-stocktake-symlinks.test.js`.** It asserts both scripts use `find -L`, then runs the `find` expression **extracted from the shipped file** rather than a retyped copy, against a 2-real + 2-linked fixture, and requires all four back by name. Extracting it means a future edit to the script cannot drift past the test. The fixture uses `fs.symlinkSync(..., 'junction')` and skips cleanly if the system refuses links.

```
  ✓ skills/skill-stocktake/scripts/scan.sh follows symlinks when enumerating
  ✓ skills/skill-stocktake/scripts/quick-diff.sh follows symlinks when enumerating
  ✓ the shipped find expression sees a symlinked skill
  Passed: 3, Failed: 0
```

Mutation-tested: reverting `scan.sh` to plain `find` fails **2 of 3** and exits 1, with `expected 2 real + 2 symlinked skills, got 2`.

**On `node tests/run-all.js`:** the full suite is not green on this machine — 53 failures in eight groups, none of which touch `skills/skill-stocktake/`, identical with and without this change (measured earlier today by running the suite on a clean `origin/main` and on a branch). I have left that box unticked rather than tick it on a suite I cannot show green; this PR adds nothing to that number.

## Known limits

`-L` follows links, so if a skills directory contains both a real directory and a symlink pointing at it, the same skill is inventoried twice under two paths. That is inherent to following links and is the correct trade against silently dropping every linked skill. `find` detects and skips symlink loops rather than spinning.

## Type of Change
- [x] `fix:` Bug fix
- [ ] `feat:` New feature
- [ ] `refactor:` Code refactoring
- [ ] `docs:` Documentation
- [ ] `test:` Tests
- [ ] `chore:` Maintenance/tooling
- [ ] `ci:` CI/CD changes

## Security & Quality Checklist
- [x] No secrets or API keys committed
- [x] JSON files validate cleanly (none changed)
- [x] Shell scripts pass shellcheck (if applicable) — `bash -n` clean on both
- [x] Pre-commit hooks pass locally (if configured)
- [x] No sensitive data exposed in logs or output
- [x] Follows conventional commits format

## If you added a skill, command, agent, hook, or CLI tool

Not applicable — no new component. One flag added to two existing scripts, plus one test file under `tests/skills/`, which `run-all.js` discovers by glob.

## Documentation
- [x] Added comments for complex logic — the comment above each `find` records why `-L` is load-bearing, so it does not get trimmed back as noise.
- [ ] README updated (if needed) — not needed

### CODE REVIEW

File: tests/skills/skill-stocktake-symlinks.test.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  write(path.join(elsewhere, 'pkg-skill'));
+  write(path.join(elsewhere, 'shared-skill'));
+  try {
+    for (const name of ['pkg-skill', 'shared-skill']) {
+      // 'junction' is the Windows form that needs no elevation; ignored elsewhere.
+      fs.symlinkSync(path.join(elsewhere, name), path.join(skills, name), 'junction');
+    }
+  } catch {
+    fs.rmSync(root, { recursive: true, force: true });
+    return null;
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply tests/skills/skill-stocktake-symlinks.test.js]: Right, and it was the worse half of the two failure modes: `runTest` counts the skip sentinel as a pass, so a broken fixture reported success.

Fixed in `902d199`. Only codes that actually mean links are unavailable — EPERM, EACCES, ENOSYS, ENOTSUP, EOPNOTSUPP — return the sentinel; anything else propagates after cleanup and fails the test. `makeFixture` also wraps the whole build so the temp root is removed on any throw, not just on a link failure.

Verified by mutation rather than by reading it: emptying `LINKS_UNSUPPORTED` turns this machine's EPERM into `✗ the shipped find expression sees linked skills` with the errno in the message, where before it was a silent pass.

File: tests/skills/skill-stocktake-symlinks.test.js [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  // Both scripts enumerate the same inventory; a quick-diff that disagreed with
+  // scan would report every symlinked skill as an addition or a removal.
+  for (const scriptPath of scripts) {
+    const rel = path.relative(repoRoot, scriptPath).split(path.sep).join('/');
+    tests.push([`${rel} follows symlinks when enumerating`, () => {
+      const line = findLine(scriptPath);
+      assert.ok(line, `no *.md find expression in ${rel}`);
+      assert.match(line, /find\s+-L\s/, `${rel} must use find -L`);
+    }]);
+  }
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply tests/skills/skill-stocktake-symlinks.test.js]: Done in `902d199` — `expressionTests` is built with `map`, the integration case is a separate array, and `main` composes them with a spread. The results are collected with `map`/`filter` too, so the pass and fail counters are no longer accumulated by mutation either.

File: tests/skills/skill-stocktake-symlinks.test.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+      const fixture = makeFixture();
+      if (fixture === null) {
+        console.log('    (skipped: this system does not allow creating links)');
+        return;
+      }
+      try {
+        // Run the exact expression the scripts ship, not a paraphrase of it.
+        const expression = findLine(scripts[0]).trim().replace(/^done < <\(/, '').replace(/\)$/, '');
+        const command = expression.replace('"$dir"', `"${toShellPath(fixture.skills)}"`);
+        const result = spawnSync(bashBinary, ['-c', command], { encoding: 'utf8' });
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply tests/skills/skill-stocktake-symlinks.test.js]: Agreed, and your version is better than a fix for the quoting alone: binding `dir` from `$1` means the shipped expression runs verbatim instead of through a `.replace()`, so the test no longer paraphrases the thing it is pinning.

Applied in `902d199`:

```js
const result = spawnSync(
  bashBinary,
  ['-c', `dir="$1"; ${expression}`, 'skill-stocktake-test', toShellPath(fixture.skills)],
  { encoding: 'utf8' }
);
```

File: tests/skills/skill-stocktake-symlinks.test.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    );
+  };
+  write(path.join(skills, 'real-a'));
+  write(path.join(skills, 'real-b'));
+  write(path.join(elsewhere, 'pkg-skill'));
+  write(path.join(elsewhere, 'shared-skill'));
+  try {
+    for (const name of ['pkg-skill', 'shared-skill']) {
+      // 'junction' is the Windows form that needs no elevation; ignored elsewhere.
+      fs.symlinkSync(path.join(elsewhere, name), path.join(skills, name), 'junction');
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply tests/skills/skill-stocktake-symlinks.test.js]: Fair — the fixture made only directory junctions, so `-type f` being false for a symlinked *file* had no coverage at all, and that is half of what the PR claims to fix.

`902d199` adds a standalone SKILL.md linked into a skill directory. It needs elevation on Windows, so rather than skipping the whole case it is reported and dropped from the expected set, and the expected list drives the count — no hardcoded number for either shape.

To be exact about what I verified: this machine cannot create file symlinks (EPERM), so the file-link assertion itself runs on CI, not here. What I did check locally is the surrounding logic, by standing a junction in for the file link — the five-entry expectation matched and no skip line was printed.

File: tests/skills/skill-stocktake-symlinks.test.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  let failed = 0;
+  for (const [name, fn] of tests) {
+    if (runTest(name, fn)) passed += 1;
+    else failed += 1;
+  }
+
+  console.log(`\n  Passed: ${passed}`);
+  console.log(`  Failed: ${failed}`);
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply tests/skills/skill-stocktake-symlinks.test.js]: Changed in `902d199`, but the stated reason does not hold and I would rather say so than leave a wrong note in the thread.

`tests/run-all.js` matches with an unanchored regex:

```js
const passedMatch = combined.match(/Passed:\s*(\d+)/);
const failedMatch = combined.match(/Failed:\s*(\d+)/);
```

Leading whitespace does not stop that from matching — `"\n  Passed: 3\n"` matches and yields `3`. The file was already being counted in the repository totals.

What does hold is the convention: of the ~200 summary lines under `tests/`, none put spaces inside the string. So the indentation is gone, on that basis.

File: skills/skill-stocktake/scripts/quick-diff.sh [hunk 6]
```diff
@@ -74,7 +74,11 @@ process_dir() {
       '{path:$path,mtime:$mtime,is_new:$is_new}' \
       > "$tmpdir/$i.json"
     i=$((i+1))
-  done < <(find "$dir" -name "*.md" -type f 2>/dev/null | sort)
+  # -L so a skill installed as a symlink is inventoried like any other. Plain
+  # find does not descend into a symlinked directory, and -type f is false for a
+  # symlinked file, so package-manager installs, app-bundled skills, and skills
+  # shared between tools were silently dropped from the count.
+  done < <(find -L "$dir" -name "*.md" -type f 2>/dev/null | sort)
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply skills/skill-stocktake/scripts/quick-diff.sh:81]: I'd rather not apply this one, and I'm leaving the thread open so you can overrule me — a target policy is your call, not mine.

The reason is that the allowlist removes the capability the change adds. #2801 is specifically about skills installed as symlinks: package-manager installs, app-bundled skills, skills shared between tools. Their canonical targets are wherever npm, Homebrew, an app bundle, or another checkout put them. I can't enumerate an approved set that covers those without it amounting to "any reachable directory", which is the thing the policy is meant to exclude.

Three things I measured rather than assumed:

**Loops terminate.** I built a cyclic junction inside a skills dir and ran the shipped expression:

```
find: File system loop detected; '/tmp/loopprobe/skills/a/self' is part of the same file system loop as '/tmp/loopprobe/skills'.
/tmp/loopprobe/skills/a/SKILL.md
exit=0
```

GNU find detects the cycle itself, reports it on stderr — which the scripts already send to `/dev/null` — and finishes with the real file still listed. No hang and no non-zero exit.

**What a found file costs.** `scan_dir_to_json` reads the `name` and `description` frontmatter fields, the mtime, and usage counts, and writes them to a local JSON inventory. Nothing is sent anywhere.

**The two roots do not carry the same trust**, and this is the part of your comment I think has real substance. `GLOBAL_DIR` is `$HOME/.claude/skills` — a link there was placed by the user or by something the user installed. `CWD_SKILLS_DIR` defaults to `$PWD/.claude/skills`, which in a freshly cloned repository is attacker-influenced: `.claude/skills/x -> /` would make the scan walk outside the tree and put paths and frontmatter of the user's own files into the report. That is a genuine delta from plain `find`, and it is information disclosure into a local report plus a large time cost, not code execution.

If you want a bound for that case, depth is the lever that does not re-break #2801, and here is the number: every one of the 399 `.md` files under this repo's own `skills/` sits at depth 2 (`<skill>/SKILL.md`) or 3 (`<skill>/references/*.md`), so `-maxdepth 3` from the skills root covers 100% of them. I have not made that change, because it also truncates non-symlinked trees and I don't know the layouts you need to keep working. Say the word and I'll add it, to both scripts, with a test.

File: tests/skills/skill-stocktake-symlinks.test.js [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+            // bound from a positional argument instead of being substituted
+            // into the command text, so a fixture path containing shell syntax
+            // is data rather than something bash parses.
+            const expression = findLine(scripts[0]).trim().replace(/^done < <\(/, '').replace(/\)$/, '');
+            const result = spawnSync(
+              bashBinary,
+              ['-c', `dir="$1"; ${expression}`, 'skill-stocktake-test', toShellPath(fixture.skills)],
+              { encoding: 'utf8' }
+            );
+            assert.strictEqual(result.status, 0, result.stderr);
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply tests/skills/skill-stocktake-symlinks.test.js:176]: Right — `spawnSync` reports a missing binary through `result.error` with `status === null`, so the exit-status assertion fired with an empty `stderr` and said nothing useful about the cause.

Fixed in `d0929e8`: ENOENT is a skip naming the binary, and any other spawn error is rethrown as itself rather than being flattened into an assertion message. Verified both ways —

```
$ ECC_TEST_BASH=__ecc_missing_bash__ node tests/skills/skill-stocktake-symlinks.test.js
    (skipped: __ecc_missing_bash__ not found)
  ✓ the shipped find expression sees linked skills
Passed: 3
Failed: 0
```

File: tests/skills/skill-stocktake-symlinks.test.js [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+              found.length,
+              fixture.expected.length,
+              `expected ${fixture.expected.join(', ')}, got ${found.length}:\n${found.join('\n')}`
+            );
+            for (const name of fixture.expected) {
+              assert.ok(
+                found.some(entry => entry.includes(`/${name}/`)),
+                `${name} missing from the inventory`
+              );
+            }
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply tests/skills/skill-stocktake-symlinks.test.js:191]: Right, and it undercut the point of the test: the two scripts disagreeing is exactly the failure this covers, so exercising one of them and static-checking the other left the gap open.

`8f08f96` maps the integration case over `scripts`, so each one's own expression is extracted and run against the fixture.

Checked by breaking only `quick-diff.sh`:

```
  ✗ skills/skill-stocktake/scripts/quick-diff.sh follows symlinks when enumerating
  ✗ skills/skill-stocktake/scripts/quick-diff.sh: the shipped find expression sees linked skills
    2 !== 4
Passed: 2
Failed: 2
```

Two failures instead of one, and the run reports 2 of 4 skills — the original #2801 symptom. `scan.sh` stays green throughout, which is what shows the cases are independent.

File: tests/skills/skill-stocktake-symlinks.test.js [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+            );
+            if (result.error) {
+              // No bash on this system: `status` is null, so the exit-status
+              // assertion below would report a confusing empty stderr instead
+              // of saying the binary is missing.
+              if (result.error.code === 'ENOENT') {
+                console.log(`    (skipped: ${bashBinary} not found)`);
+                return;
+              }
+              throw result.error;
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply tests/skills/skill-stocktake-symlinks.test.js:174]: Agreed — and it is the sharper version of the change you asked for last round. An absent platform feature and a misconfigured one are different things, and only the first should be quiet.

Applied verbatim in `51a319b`. Verified in both directions:

```
$ ECC_TEST_BASH=__ecc_missing_bash__ node tests/skills/skill-stocktake-symlinks.test.js
  ✗ skills/skill-stocktake/scripts/scan.sh: the shipped find expression sees linked skills
    spawnSync __ecc_missing_bash__ ENOENT
Passed: 2
Failed: 2

$ ECC_TEST_BASH=bash node tests/skills/skill-stocktake-symlinks.test.js
Passed: 4
Failed: 0
```
